File: src/main/java/org/bakkes/game/controller/state/battle/BattleStateModule.java
package org.bakkes.game.controller.state.battle;

import java.util.Collection;
import java.util.LinkedList;

import org.bakkes.game.AModule;
import org.bakkes.game.controller.event.input.BattleEndListner;
import org.bakkes.game.controller.event.input.IKeyListener;
import org.bakkes.game.model.map.IAreaNameAcces;
import org.newdawn.slick.state.StateBasedGame;

import com.google.inject.Inject;
import com.google.inject.Provider;
import com.google.inject.Provides;

/**
 * bridging module
 */
public class BattleStateModule extends AModule{
	private @Inject Provider<StateBasedGame> game;
    private @Inject Provider<IAreaNameAcces> named;
    @Override
	public void configure(){
    }
	@Provides
	StateBasedGame provideGame(){
		return game.get();
	}
	@Provides
	IAreaNameAcces provideAcces(){
		return named.get();
	}
	@Provides
    public Collection<IKeyListener> providesKeyListeners(
    		final LinkedList<IKeyListener> result,
    		final BattleEndListner listener
        ){
    	result.add(listener);
    	return result;
    }
}

File: src/main/scala/org/bakkes/game/controller/state/battle/BattleStateModule.scala
package org.bakkes.game.controller.state.battle;

import com.google.inject.name.{Named, Names}
import com.google.inject.{Provider, Provides, Singleton}
import net.codingwell.scalaguice.ScalaModule
import org.bakkes.game.AModule
import org.bakkes.game.controller.MenuController
import org.bakkes.game.controller.state.event.input.CompositeKeyListener
import org.bakkes.game.controller.state.battle.event.BattleMenuHandler
import org.bakkes.game.controller.state.battle.event.key.BattleEndListner
import org.bakkes.game.model.pokemon.Pokemon
import org.bakkes.game.view.components.{ITextableShape, Menu}

/**
 * bridging module
 */
class BattleStateModule extends AModule with ScalaModule{
	override def configure{
    	//bind(Menu.class).annotatedWith(Names.named("moves")).to(Menu.class).in(Singleton.class);
    	bind[BattleLoader].in[Singleton]
    	bind[Menu].annotatedWith(Names.named("moves")).to[Menu].in[Singleton]
    }

    @Provides
    @Named("current players")
    def provideCurrentPlayerPokemon (loader:BattleLoader) : Pokemon = loader.getCurrentPlayerPokemon();

    @Provides
    @Named("current enemys")
    def provideCurrentEnemyPokemon (loader:BattleLoader) : Pokemon = loader.getCurrentEnemyPokemon();
	@Provides
	@Named("battle")
	def provideBattleMenu(result:Menu, lineProvider:Provider[ITextableShape]) : Menu = {
		return result
	}

    @Provides
    @Named("battle")
    def provideMenuController(
    		@Named("moves") moveMenu:Menu, 
    		result:MenuController, 
    		handler:BattleMenuHandler
    ) : MenuController = {
    	moveMenu.clear()
    	result.set(moveMenu,handler)
    	return result
    }

    @Provides
    @Named("battle")
    def provideKeyInput(
    	result:CompositeKeyListener,
    	end:BattleEndListner,
    	@Named("battle") controller:MenuController
    ):CompositeKeyListener = {
    	result.clear()
    	result.add(end)
    	result.add(controller)
    	return result
    }
    @Provides
    @Named("current")
    def provideBattle(loader:BattleLoader) : Battle = loader.getCurrentBattle();
}
